packages/query-persist-client-core/src/createPersister.ts

```diff
@@ -0,0 +1,161 @@
+import type {
+  Query,
+  QueryFunctionContext,
+  QueryKey,
+  QueryState,
+} from '@tanstack/query-core'
+
+export interface PersistedQuery {
+  buster: string
+  queryHash: string
+  queryKey: QueryKey
+  state: QueryState
+}
+
+export interface AsyncStorage {
+  getItem: (key: string) => Promise<string | undefined | null>
+  setItem: (key: string, value: string) => Promise<unknown>
+  removeItem: (key: string) => Promise<void>
+}
+
+export interface StoragePersisterOptions {
+  /** The storage client used for setting and retrieving items from cache.
+   * For SSR pass in `undefined`.
+   */
+  storage: AsyncStorage | Storage | undefined | null
+  /**
+   * How to serialize the data to storage.
+   * @default `JSON.stringify`
+   */
+  serialize?: (persistedQuery: PersistedQuery) => string
+  /**
+   * How to deserialize the data from storage.
+   * @default `JSON.parse`
+   */
+  deserialize?: (cachedString: string) => PersistedQuery
+  /**
+   * A unique string that can be used to forcefully invalidate existing caches,
+   * if they do not share the same buster string
+   */
+  buster?: string
+  /**
+   * The max-allowed age of the cache in milliseconds.
+   * If a persisted cache is found that is older than this
+   * time, it will be discarded
+   * @default 24 hours
+   */
+  maxAge?: number
+  /**
+   * Prefix to be used for storage key.
+   * Storage key is a combination of prefix and query hash in a form of `prefix-queryHash`.
+   * @default 'tanstack-query'
+   */
+  prefix?: string
+  /**
+   * Filter function returning whether current query should be run trough persister.
+   */
+  queryFilter?: (queryKey: QueryKey, query: Query) => boolean
```

Comment:
@TkDodo Since `experimental_createPersister` already returns `persister` function it would be hard to add predicate option, so instead i have added `queryFilter` as an option that will be evaluated before `restoring` or `persisting`

Reply:
that's a good idea, but why not support full QueryFilters?

https://github.com/TanStack/query/blob/13b17ee45de3299dd3d3a2dfc3d23ad2f024882a/packages/query-core/src/utils.ts#L13

Reply:
what do you think?

https://github.com/TanStack/query/pull/5125/commits/36112eaa84714b1cfc1493a9f152d7282df34c6c

Reply:
Nice 👍